wagtail/snippets/models.py

```diff
@@ -42,49 +52,57 @@ def get_edit_url(self, instance):
 
 
 def register_snippet(model, viewset=None):
-    if model not in SNIPPET_MODELS:
-        from wagtail.snippets.views.chooser import SnippetChooserViewSet
-        from wagtail.snippets.views.snippets import SnippetViewSet
-
-        model.get_usage = lambda obj: ReferenceIndex.get_references_to(
-            obj
-        ).group_by_source_object()
-        model.usage_url = get_snippet_usage_url
-        model.get_admin_base_path = get_admin_base_path
-        model.get_admin_url_namespace = get_admin_url_namespace
-
-        if viewset is None:
-            viewset = SnippetViewSet
-        elif isinstance(viewset, str):
-            viewset = import_string(viewset)
-
-        admin_viewset = viewset(
-            model.get_admin_url_namespace(),
-            model=model,
-            url_prefix=model.get_admin_base_path(),
-        )
-
-        chooser_viewset = SnippetChooserViewSet(
-            f"wagtailsnippetchoosers_{model._meta.app_label}_{model._meta.model_name}",
-            model=model,
-            url_prefix=f"snippets/choose/{model._meta.app_label}/{model._meta.model_name}",
-        )
-
-        viewsets.register(admin_viewset)
-        viewsets.register(chooser_viewset)
-
-        SNIPPET_MODELS.append(model)
-        SNIPPET_MODELS.sort(key=lambda x: x._meta.verbose_name)
-
-        @checks.register("panels")
-        def modeladmin_model_check(app_configs, **kwargs):
-            errors = check_panels_in_model(model, "snippets")
-            return errors
-
-        # Set up admin model forms to use AdminSnippetChooser for any ForeignKey to this model
-        register_form_field_override(
-            ForeignKey, to=model, override={"widget": AdminSnippetChooser(model=model)}
-        )
+    if DEFER_REGISTRATION:
+        # Models may not have been fully loaded yet, so defer registration until they are -
+        # add it to the list of registrations to be processed by register_deferred_snippets
+        DEFERRED_REGISTRATIONS.append((model, viewset))
+        return model
+    elif model in SNIPPET_MODELS:
```

Comment:
For clarity, would it be worth breaking everything from here downwards into a separate `register_snippet_deferred()` function, and calling that from `register_deferred_snippets()`?

Reply:
I think we might be able to do away with `DEFER_REGISTRATION` in that case, and always have `register_snippet()` defer until later

Reply:
Okay sure. Happy with that. I originally thought (before my second comment) that `register_snippet` could also call `_register_snippet_actual`, and that would put slightly more distance between the problematic code. However, if the problem is resolved by the conditional, I suppose there's not much to be gained.

Reply:
@ababic I guess the gist of your suggestion is that calling `register_snippet` again from `register_deferred_snippets` (and thus re-running the 'defer or not' check) makes things a bit too "circular", and that's a fair point. Have now shuffled things in 07ae39d152815dcb523ab06949defc52002555c5 - see what you think.

(I have a bit of a hang-up about using the word "actual" here, as it implies a sort of untrustworthiness about the function that doesn't have it :-) Hopefully "immediate" works as a replacement.)

Reply:
@gasman My reasoning was probably a bit more 'misunderstanding driven' than what you describe, but this nails it. Thank you!